src/core_ocean/shared/mpas_ocn_tracer_hmix_redi.F

```diff
@@ -165,285 +166,351 @@ subroutine ocn_tracer_hmix_redi_tend(meshPool, scratchPool, layerThicknessEdge,
+               ! 2.0 in next line is because a dot product on a C-grid
+               ! requires a factor of 1/2 to average to the cell center.
+               flux_term3 = RediKappa(iCell) * 2.0_RKIND * &
+                   (kappaRediSfcTaper(k,iCell)*kappaGMcell(k,iCell)*fluxRediZTop(iTr,k) - &
+                    kappaRediSfcTaper(k+1,iCell)*kappaGMcell(k+1,iCell)*fluxRediZTop(iTr,k+1))
+
+               redi_term3(iTr,k,iCell) = redi_term3(iTr,k,iCell) + flux_term3
             end do
          end do
-         !$omp end do
-
-         if(.not.config_disable_redi_horizontal_term3) then
-            nCells = nCellsArray( 1 )
-            !$omp do schedule(runtime)
-            do iCell = 1, nCells
-               ! impose no-flux boundary conditions at top and bottom of column
-               gradHTracerSlopedTopOfCell(1,iCell) = 0.0_RKIND
-               gradHTracerSlopedTopOfCell(maxLevelCell(iCell)+1,iCell) = 0.0_RKIND
-               do k = 1, maxLevelCell(iCell)
-                  tend(iTracer,k,iCell) = tend(iTracer,k,iCell) + config_Redi_kappa * &
-                      (gradHTracerSlopedTopOfCell(k,iCell) - gradHTracerSlopedTopOfCell(k+1,iCell))
-               end do
-            end do
-            !$omp end do
-         endif
-
-      end do  ! iTracer
-
-      call mpas_deallocate_scratch_field(gradTracerEdgeField, .true.)
-      call mpas_deallocate_scratch_field(gradTracerTopOfEdgeField, .true.)
-      call mpas_deallocate_scratch_field(gradHTracerSlopedTopOfCellField, .true.)
-      call mpas_deallocate_scratch_field(dTracerdZTopOfCellField, .true.)
-      call mpas_deallocate_scratch_field(dTracerdZTopOfEdgeField, .true.)
-      call mpas_deallocate_scratch_field(areaCellSumField, .true.)
+      end do !loop over cells
+      !  !$omp end do
+      deallocate(fluxRediZTop)
+
+      !loop over cells and check for out of bounds
+      if(isActiveTracer) then
+        minimumVal(1) = -2.0_RKIND
+        minimumVal(2:nTracers) = 0.0_RKIND
+        rediLimiterCount(:,:) = 0
+      else
+        minimumVal(:) = 0.0_RKIND
+      endif
 
-      call mpas_timer_stop("tracer redi")
+      !  !$omp do schedule(runtime) private(k, iTr, tempTracer, iEdge)
+      do iCell = 1, nCells
+        do k = 1, maxLevelCell(iCell)
+           do iTr=1,ntracers
+              tempTracer = tracers(iTr,k,iCell) + dt*(redi_term1(iTr,k,iCell) + redi_term2(iTr,k,iCell) + &
+                              redi_term3(iTr,k,iCell))/layerThickness(k,iCell)
+
+              if(tempTracer < minimumVal(iTr)) then
+
+                  do i = 1, nEdgesOnCell(iCell)
+                     iEdge = edgesOnCell(i, iCell)
+                     redi_term2_edge(iTr,k,iEdge) = 0.0_RKIND
+                  enddo
+
+                  redi_term3_topOfCell(iTr,k,iCell) = 0.0_RKIND
+                  redi_term3_topOfCell(iTr,k+1,iCell) = 0.0_RKIND
+                  rediLimiterCount(k,iCell) = 1.0_RKIND
```

Comment:
@vanroekel this line sets the counter to 1 for any tracer group, but you set rediLimiterCount to 0 above only for active tracers. You just want it for active tracers, right? We can also increment in case T&S are both reset:
`rediLimiterCount(k,iCell) = rediLimiterCount(k,iCell) + 1`

Reply:
oh, yes, definitely just active tracers, I am not sure if incrementing would work well.  There could be times when salinity is reset and not temperature and vice versa and so rediLimiterCount=1 wouldn't tell us which, but = 2 would say both.  I would say we only do it for temperature, or I could go back and make that part of tracer groups.